include/odp/api/spec/packet_io.h

```diff
@@ -512,18 +512,38 @@ typedef struct odp_pktio_config_t {
 	 *
 	 *  Enable/disable inline outbound IPSEC operation. When enabled IPSEC
 	 *  outbound processing can send outgoing IPSEC packets directly
-	 *  to the pktio interface for output. IPSEC configuration is done
+	 *  to the pktio interface queue for output. IPSEC configuration is done
 	 *  through the IPSEC API.
 	 *
-	 *  Outbound IPSEC inline operation cannot be combined with traffic
-	 *  manager (ODP_PKTOUT_MODE_TM).
+	 *  When enabled, odp_ipsec_out_inline() would use packet out queue for
+	 *  packet transmission.
 	 *
-	 *  0: Disable outbound IPSEC inline operation (default)
-	 *  1: Enable outbound IPSEC inline operation
+	 * * 0: Disable outbound IPSEC inline operation (default)
+	 * * 1: Enable outbound IPSEC inline operation when
+	 *     outbound_ipsec_tm_queue = 0
 	 *
 	 *  @see odp_ipsec_config(), odp_ipsec_sa_create()
+	 *  odp_ipsec_out_inline()
 	 */
-	odp_bool_t outbound_ipsec;
+	odp_bool_t outbound_ipsec_pktout_queue;
+
+	/** Outbound IPSEC inlined with traffic manager(TM)
+	 *
+	 *  Enable/disable inline outbound IPSEC operation with traffic manager
+	 *  queue. When enabled IPSEC outbound processing can submit outgoing
+	 *  packets to traffic manager subsystem. IPSEC configuration is done
+	 *  through the IPSEC API.
+	 *
+	 *  When enabled, odp_ipsec_out_inline() would use TM queue for packet
+	 *  transmission.
+	 *
+	 * * 0: Disable outbound IPSEC inline operation with TM queues (default)
+	 * * 1: Enable outbound IPSEC inline operation with TM queues
+	 *
+	 *  @see odp_ipsec_config(), odp_ipsec_sa_create(),
+	 *  odp_ipsec_out_inline()
+	 */
+	odp_bool_t outbound_ipsec_tm_queue;
```

Comment:
Why does this need to be introduced instead of relying on the out_mode (e.g. ODP_PKTOUT_MODE) configured for the pktout? It would seem simpler and more clear that pktout either uses or does not use TM based on the existing mode config and inline IPsec just follows that.

This additional setting would seem to allow one to configure the use or non-use of TM separately for normal output operations and IPsec inline output operations. I am not sure if that is a good thing. It may make implementation harder in some environments.

Reply:
That should be fine. Thanks for the suggestion. So what you propose is 

if (odp_pktio_param_t.out_mode == ODP_PACKET_MODE_TM) {
    /* Use TM queues */
} else {
    /* Use pktout */
}

Above should be fine. 

To answer your other question,

> Why is pktout queue parameter needed, i.e. what problem does it solve?

Because for a regular lookaside mode of operation, the packet being transmitted after IPsec can be at pktout queue level. With inline that will not be possible and any application which is trying to use the pktout queues for distribution may find that missing.